### PR TITLE
Add implicit PackageTargetFallback for net461 for netcoreapp and netstandard 2.0

### PR DESCRIPTION
**Pull request by livarcocc**
Fixes https://github.com/dotnet/sdk/issues/1025

@dotnet/dotnet-cli

### CODE REVIEW

**Comment by gulbanana** [comment]
So if i want to build netstandard 2.0 libraries which don't risk accidentally referencing netfx packages, would it be correct to set property DisableImplicitPackageTargetFallback to True?

**Comment by livarcocc** [comment]
@gulbanana yes. also, unless you explicitly turned it off, when ptf is used, you will get a warning.

**Comment by livarcocc** [comment]
@dotnet-bot Test Windows_NT_FullFramework Debug

**Comment by livarcocc** [comment]
@dsplaisted @nguerrera PTAL

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyNuGetReferenceCompat.cs [hunk 1]
```diff
@@ -180,6 +141,140 @@ public class GivenThatWeWantToVerifyNuGetReferenceCompat : SdkTest, IClassFixtur
             }
         }
 
+        [WindowsOnlyTheoryAttribute]
+        [InlineData("netstandard2.0")]
+        [InlineData("netcoreapp2.0")]
+        public void Net461_is_implicit_for_Netstandard_and_Netcore_20(string targetFramework)
+        {
+            if (UsingFullFrameworkMSBuild)
```
Comment (nguerrera) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyNuGetReferenceCompat.cs]: While we're getting better at xunit, should we have some sort of attribute to indicate core msbuild only?

File: test/Microsoft.NET.TestFramework/ProjectConstruction/TestProject.cs [hunk 2]
```diff
@@ -168,6 +170,11 @@ internal void Create(TestAsset targetTestAsset, string testProjectsSourceFolder)
                 targetFrameworkVersionElement.SetValue(this.TargetFrameworkVersion);
             }
 
+            foreach (var additionalProperty in AdditionalProperties)
+            {
+                propertyGroup.Add(new XElement($"{ns}{additionalProperty.Key}", additionalProperty.Value));
```
Comment (nguerrera) [inline test/Microsoft.NET.TestFramework/ProjectConstruction/TestProject.cs]: This should be ns + additionalProperty.Key like the others above.

You want https://msdn.microsoft.com/en-us/library/system.xml.linq.xnamespace.op_addition(v=vs.110).aspx, not string concatenation.  I suspect this is only working now because the projects under test have no namespace declared.

**Review by nguerrera — APPROVED** [review]

File: build/build.proj [hunk 3]
```diff
@@ -170,12 +170,7 @@
           LogStandardErrorAsError="true"
           WorkingDirectory="$(TestsDirectory)"
           IgnoreExitCode="true"
-          >
-      <Output TaskParameter="ExitCode" PropertyName="ExitCode" />
-    </Exec>
-
-    <Error Text="There were test failures, for full log see %(XmlTestFile.FullPath)" Condition="$(ExitCode) != 0" />
-
```
Comment (livarcocc) [inline build/build.proj]: Removed it because xunit returns non-zero even when we have skipped tests. Which now that we have skip attributes happened and caused the build to fail. We don't need this anymore because we are parsing the testResult.xml file and jenkins will fail if there are any failed tests in there.

File: test/Microsoft.NET.Build.Tests/DeleteNuGetArtifactsFixture.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                }
+                //  Delete the generated NuGet packages in the cache.
+                foreach (string dir in Directory.EnumerateDirectories(RepoInfo.NuGetCachePath, ConstantStringValues.DependencyDirectoryNamePrefix + "*"))
+                {
+                    Directory.Delete(dir, true);
+                }
+            }
+            catch
+            {
+                // No-Op; as this is a precaution - do not throw an exception.
```
Comment (livarcocc) [inline test/Microsoft.NET.Build.Tests/DeleteNuGetArtifactsFixture.cs]: @johnbeisner can you help clarify this?

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyNuGetReferenceCompat.cs [hunk 5]
```diff
@@ -180,6 +141,140 @@ public class GivenThatWeWantToVerifyNuGetReferenceCompat : SdkTest, IClassFixtur
             }
         }
 
+        [WindowsOnlyTheoryAttribute]
```
Comment (livarcocc) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyNuGetReferenceCompat.cs]: Oops. Copy/paste. Will do.

File: test/Microsoft.NET.TestFramework/ProjectConstruction/TestProject.cs (hunk 2 above)
Comment (livarcocc) [inline test/Microsoft.NET.TestFramework/ProjectConstruction/TestProject.cs]: Fixed.

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyNuGetReferenceCompat.cs [hunk 6]
```diff
@@ -180,6 +141,140 @@ public class GivenThatWeWantToVerifyNuGetReferenceCompat : SdkTest, IClassFixtur
...
+        [InlineData("netstandard2.0")]
+        [InlineData("netcoreapp2.0")]
+        public void Net461_is_implicit_for_Netstandard_and_Netcore_20(string targetFramework)
```
Comment (dsplaisted) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyNuGetReferenceCompat.cs]: I wouldn't include the values of the theory parameters in the name of the test.  I'd call it something like `Net461_is_implicit_package_target_fallback`.

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyNuGetReferenceCompat.cs [hunk 7]
```diff
@@ -180,6 +141,140 @@ public class GivenThatWeWantToVerifyNuGetReferenceCompat : SdkTest, IClassFixtur
...
+            restoreCommand.AddSource(Path.GetDirectoryName(_net461PackageReference.NupkgPath));
+            restoreCommand.Execute().Should().Fail();
+        }
+
+        private TestAsset CreateTestAsset(
+            string testProjectName,
+            string targetFramework,
+            Dictionary<string, string> additionalProperties = null)
+        {
+            _net461PackageReference = CreateNet461Package();
```
Comment (dsplaisted) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyNuGetReferenceCompat.cs]: It feels messy to use a field of the class to pass this object around.  I think I'm used to tests not storing any instance state.

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyNuGetReferenceCompat.cs [hunk 8]
```diff
@@ -180,6 +141,140 @@ public class GivenThatWeWantToVerifyNuGetReferenceCompat : SdkTest, IClassFixtur
...
+
+            return testProjectTestAsset;
+        }
+
+        private TestPackageReference CreateNet461Package()
+        {
+            var net461Project = 
+                new TestProject
+                {
+                    Name = $"net461_pkg",
```
Comment (dsplaisted) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyNuGetReferenceCompat.cs]: Nit: no need for `$` in front of string here.

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyNuGetReferenceCompat.cs [hunk 9]
```diff
@@ -180,6 +141,140 @@ public class GivenThatWeWantToVerifyNuGetReferenceCompat : SdkTest, IClassFixtur
...
+
+            var testProjectTestAsset = CreateTestAsset(testProjectName, targetFramework);
+
+            var restoreCommand = testProjectTestAsset.GetRestoreCommand(relativePath: testProjectName);
+            restoreCommand.AddSource(Path.GetDirectoryName(_net461PackageReference.NupkgPath));
+            restoreCommand.Execute().Should().Fail();
+        }
+
+        [WindowsOnlyFact]
+        public void It_is_possible_to_disabled_net461_implicit_package_target_fallback()
```
Comment (dsplaisted) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyNuGetReferenceCompat.cs]: Nit: change `disabled` to `disable` in the method name.

**Review by dsplaisted — APPROVED** [review]

File: test/Microsoft.NET.TestFramework/CoreMSBuildAndWindowsOnlyTheoryAttribute.cs [hunk 10]
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using Microsoft.DotNet.PlatformAbstractions;
+using Xunit;
+
+namespace Microsoft.NET.TestFramework
+{
+    public class CoreMSBuildAndWindowsOnlyTheoryAttribute : TheoryAttribute
```
Comment (nguerrera) [inline test/Microsoft.NET.TestFramework/CoreMSBuildAndWindowsOnlyTheoryAttribute.cs]: Roslyn has a nice pattern for combining common skip conditions: https://github.com/dotnet/roslyn/blob/master/src/Test/Utilities/Portable/Assert/ConditionalFactAttribute.cs

I think we should steal it as it allows composing things like WindowsOnly and CoreMSBuildOnly and also has the benefit of consistency with other dotnet repos.

Comment (livarcocc) [inline test/Microsoft.NET.TestFramework/CoreMSBuildAndWindowsOnlyTheoryAttribute.cs]: Can we do that in a separate PR though? I would like to merge this one and then we can bring that over.

I agree that's much better.
